core/src/main/java/com/linecorp/armeria/client/endpoint/WeightBasedRandomEndpointSelector.java

```diff
@@ -0,0 +1,156 @@
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.client.endpoint;
+
+import java.util.List;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;
+
+import javax.annotation.Nullable;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableList.Builder;
+
+import com.linecorp.armeria.client.Endpoint;
+
+final class WeightBasedRandomEndpointSelector {
+    private static final AtomicIntegerFieldUpdater<WeightBasedRandomEndpointSelector> removingEntryUpdater =
+            AtomicIntegerFieldUpdater.newUpdater(WeightBasedRandomEndpointSelector.class, "removingEntry");
+
+    private final List<Entry> entries;
+    private final long totalWeight;
+    private final List<Entry> currentEntries;
+    private long currentTotalWeight;
+
+    private volatile int removingEntry;
+
+    WeightBasedRandomEndpointSelector(List<Endpoint> endpoints) {
+        final Builder<Entry> builder = ImmutableList.builder();
+
+        long totalWeight = 0;
+        for (Endpoint endpoint : endpoints) {
+            builder.add(new Entry(endpoint, totalWeight));
+            totalWeight += endpoint.weight();
+        }
+        this.totalWeight = totalWeight;
+        currentTotalWeight = totalWeight;
+        entries = builder.build();
+        currentEntries = new CopyOnWriteArrayList<>(entries);
+    }
+
+    @VisibleForTesting
+    List<Entry> entries() {
+        return entries;
+    }
+
+    @Nullable
+    Endpoint selectEndpoint() {
+        if (entries.isEmpty()) {
+            return null;
+        }
+        for (;;) {
+            final Endpoint endpoint = selectEndpoint0();
+            if (endpoint != null) {
+                return endpoint;
+            }
+        }
+    }
+
+    @Nullable
+    Endpoint selectEndpoint0() {
+        final long nextLong = ThreadLocalRandom.current().nextLong(currentTotalWeight);
+        // There's a chance that currentTotalWeight is changed before looping currentEntries.
+        // However, we have counters and choosing an endpoint doesn't have to be exact so no big deal.
+        // TODO(minwoox): Use binary search when the number of endpoints is greater than N.
+        for (Entry entry : currentEntries) {
+            if (entry.lowerBound >= nextLong) {
+                if (entry.increaseCounter()) {
+                    return entry.endpoint();
+                }
+                // Should remove the entry from currentEntries.
+                removeEntry(entry);
+                return null;
+            }
+        }
+
+        return null;
+    }
+
+    private void removeEntry(Entry entry) {
+        for (;;) {
+            if (removingEntryUpdater.compareAndSet(this, 0, 1)) {
```

Comment:
This looks like it's using an intentional spin-lock instead of just a synchronized block (in many code pieces, we have a loop + atomic to prevent duplicate updates of a field, where it does make sense since the loop ends when another thread does the update, here we're literally just looping until the lock opens). FWIU, Java does spin a bit before sleeping a thread on locks so it's generally not recommended to use a spin lock in Java code.

Reply:
Yeah, let me use the `synchronized` block when the `currentEntries` is modified. 😉